File: 백준/Gold/17089. 세 친구/세 친구.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

    public static void main(String[] args) throws IOException {
        // 0. 빠른 입력을 위한 설정
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        // 1. 데이터 저장 및 준비 단계
        // 친구 관계를 빠르게 확인하기 위한 인접 행렬
        boolean[][] areFriends = new boolean[N + 1][N + 1];
        // 각 사람의 친구 수를 저장하기 위한 배열
        int[] friendCount = new int[N + 1];

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());

            // 양방향 친구 관계 설정
            areFriends[u][v] = true;
            areFriends[v][u] = true;

            // 각자의 친구 수 증가
            friendCount[u]++;
            friendCount[v]++;
        }

        // 최솟값을 저장할 변수. 충분히 큰 값으로 초기화
        int minSum = Integer.MAX_VALUE;

        // 2. 탐색 및 계산 단계
        // 모든 사람 A, B, C의 조합을 확인
        for (int A = 1; A <= N; A++) {
            for (int B = A + 1; B <= N; B++) {
                // 먼저 A와 B가 친구인지 확인 (이것이 핵심!)
                if (areFriends[A][B]) {
                    // A와 B가 친구라면, 이 둘과 모두 친구인 C를 찾음
                    for (int C = B + 1; C <= N; C++) {
                        // C가 A와도 친구이고, B와도 친구인지 확인
                        if (areFriends[A][C] && areFriends[B][C]) {
                            // 세 친구(A, B, C)를 찾았음!
                            // 문제의 조건에 따라 친구 수의 합을 계산
                            int currentSum = (friendCount[A] - 2) + (friendCount[B] - 2) + (friendCount[C] - 2);
                            
                            // 현재 합계가 이전에 찾은 최솟값보다 작으면 갱신
                            minSum = Math.min(minSum, currentSum);
                        }
                    }
                }
            }
        }

        // 3. 결과 출력 단계
        // minSum이 초기값 그대로라면 세 친구 관계를 찾지 못한 것
        if (minSum == Integer.MAX_VALUE) {
            System.out.println(-1);
        } else {
            System.out.println(minSum);
        }
    }
}

File: 백준/Gold/17089. 세 친구/세 친구.py
import sys

def solve():
    """
    Finds the minimum sum of friends for a group of three mutual friends.
    """
    # 0. Fast input setup
    input = sys.stdin.readline
    
    try:
        n, m = map(int, input().split())
    except ValueError:
        # Handles empty input lines that might occur
        return

    # 1. Prepare data structures
    # Adjacency matrix for quick O(1) friend lookup
    are_friends = [[False] * (n + 1) for _ in range(n + 1)]
    # Array to store the total friend count for each person
    friend_count = [0] * (n + 1)

    for _ in range(m):
        u, v = map(int, input().split())
        
        # Set bidirectional friendship
        are_friends[u][v] = True
        are_friends[v][u] = True
        
        # Increment friend count for both
        friend_count[u] += 1
        friend_count[v] += 1
        
    # Variable to store the minimum sum, initialized to a very large value
    min_sum = sys.maxsize

    # 2. Search and calculate
    # Iterate through all combinations of three people (A, B, C)
    for a in range(1, n + 1):
        for b in range(a + 1, n + 1):
            # Optimization: only proceed if A and B are friends
            if are_friends[a][b]:
                for c in range(b + 1, n + 1):
                    # Check if C is friends with both A and B to form a triangle
                    if are_friends[a][c] and are_friends[b][c]:
                        # A, B, and C are mutual friends.
                        # Calculate the sum of their friends *outside* this group.
                        # For each person, subtract the other 2 friends in the group.
                        current_sum = (friend_count[a] - 2) + (friend_count[b] - 2) + (friend_count[c] - 2)
                        
                        # Update the minimum sum if the current one is smaller
                        min_sum = min(min_sum, current_sum)

    # 3. Print the result
    # If min_sum was never updated, no group of three was found
    if min_sum == sys.maxsize:
        print(-1)
    else:
        print(min_sum)

solve()